### PR TITLE
Allow to launch server without Redis

### PR DESCRIPTION
**Pull request by jrmi**
This PR propose a way to have multiple store backend to avoid Redis use on small instance or in dev mode.

I also enhance a little the Readme and install process.

### CODE REVIEW

File: server/src/index.js [hunk 1]
```diff
@@ -106,52 +121,52 @@ const protocol = (process.env.PROTOCOL || 'http') === 'http' ? http : https;
...
 const io = Io(server, {
   pingInterval: 20000,
-  pingTimeout: 5000
+  pingTimeout: 5000,
 });
-io.adapter(socketRedis(process.env.REDIS_URL));
+
+// Only use redis if we have a REDIS_URL in config
+if (process.env.REDIS_URL !== undefined) {
+  io.adapter(socketRedis(process.env.REDIS_URL));
```
Comment (seripap) [inline server/src/index.js:129]: Maybe we initializer `io` before creating a `store`, then we can do this portion on the store setup

File: server/src/utils/store.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import Redis from 'redis';
```
Comment (seripap) [inline server/src/utils/store.js:1]: This is fine for now- eventually we'd probably want to break this out into their individual files and have `store.js` just be the exporter for `Redis` and `Memory`. Even better we'd move the entire store class outside of utils into its own root folder.

File: server/src/index.js (hunk 1 above)
Comment (seripap) [reply server/src/index.js:129]: After looking below, we'd want our store to know that it's using a custom io adapter so we'd want to make that flag readily available for use, like setting a `store.hasAdapter = bool` flag after initializing the adapter.

File: server/src/socket.js [hunk 3]
```diff
@@ -32,30 +32,41 @@ export default class Socket {
...
-  joinRoom(roomId, socketId) {
+  joinRoom(roomId, socket) {
     return new Promise((resolve, reject) => {
-      getIO().of('/').adapter.remoteJoin(socketId, roomId, (err) => {
-        if (err) {
-          reject()
-        }
-        resolve()
-      });
+      if (process.env.REDIS_URL !== undefined) {
```
Comment (seripap) [inline server/src/socket.js:53]: Let's extend our store class to see if an adapter was initialized so we're not assuming that if a `STORE_HOST` is set, it is available. Then we are able to do `getStore().hasAdapter` and make decisions from there.

File: server/.env.dist [hunk 4]
```diff
@@ -5,4 +5,5 @@ ABUSE_FROM_EMAIL_ADDRESS=Darkwire <no-reply@darkwire.io>
 REDIS_URL=redis://localhost:6379
```
[the remark this answers is not in the document]
Comment (jrmi) [reply server/.env.dist]: Done

File: server/src/index.js (hunk 1 above)
Comment (jrmi) [reply server/src/index.js]: I've added the boolean `hasSocketAdapter` property and a `getSocketAdapter` to creat the adapter on demand.

File: server/src/utils/store.js (hunk 2 above)
Comment (jrmi) [reply server/src/utils/store.js]: Their is now a file for each in the store folder.

File: readme.md [hunk 5]
```diff
@@ -16,6 +16,22 @@ The Darkwire.io [web client](/client) is written in JavaScript with React JS and
 
 ### Development
 
+#### Prerequisites
+
+Copy `.env.dist` files in `server/` and `client/` directoies without the `.dist`
```
Comment (alanfriedman) [inline readme.md:21]:
```suggestion
Copy `.env.dist` files in `server/` and `client/` directories without the `.dist`
```

File: server/src/index.js [hunk 6]
```diff
@@ -1,27 +1,34 @@
...
-
-bluebird.promisifyAll(Redis.RedisClient.prototype);
-bluebird.promisifyAll(Redis.Multi.prototype);
-
-const redis = Redis.createClient(process.env.REDIS_URL)
+import { pollForInactiveRooms } from './inactive_rooms';
+import RedisStore from './store/Redis';
+import MemoryStore from './store/Memory';
+
+const storeBackend = process.env.STORE_BACKEND || 'redis';
```
Comment (alanfriedman) [inline server/src/index.js:18]: Just to keep this initialization file tidy, would you mind moving this, the switch/case and `getStore` into a separate module? Then we can just have `const store = getStore()` here

File: server/src/store/Memory.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ */
+export class MemoryStore {
+  constructor() {
+    this.store = {};
+    this.hasSocketAdapter = false;
+  }
+
+  async get(key, field) {
+    if (this.store[key] === undefined || this.store[key][field] === undefined) {
+      return '';
```
Comment (alanfriedman) [inline server/src/store/Memory.js:12]: For consistency with Redis, this should probably [return null](https://redis.io/commands/hget) if the field doesn't exist.
```suggestion
      return null;
```

File: server/src/store/Memory.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  async set(key, field, value) {
+    if (this.store[key] === undefined) {
+      this.store[key] = {};
+    }
+    this.store[key][field] = value;
+  }
+
+  async del(key, field) {
+    if (this.store[key] === undefined || this.store[key][field] === undefined) {
+      return;
```
Comment (alanfriedman) [inline server/src/store/Memory.js:34]: [Redis returns 0](https://redis.io/commands/hdel) when the field doesn't exist. Do we want to match the Redis API in our stores, or define our own store API? I'm fine with either as long as it's consistent across stores.
```suggestion
      return 0;
```

File: server/src/store/Memory.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    }
+    delete this.store[key][field];
```
Comment (alanfriedman) [inline server/src/store/Memory.js:36]: Redis [returns the number of fields removed](https://redis.io/commands/hdel) from the hash. We could just return `1` upon success.
```suggestion
    delete this.store[key][field];
    return 1;
```

File: server/src/store/Memory.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  async inc(key, field, inc = 1) {
+    this.store[key][field] += inc;
```
Comment (alanfriedman) [inline server/src/store/Memory.js:40]: To [match redis](https://redis.io/commands/incrby)
```suggestion
    this.store[key][field] += inc;
    return this.store[key][field];
```

File: server/src/store/Memory.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    return this.store[key];
+  }
+
+  async set(key, field, value) {
+    if (this.store[key] === undefined) {
+      this.store[key] = {};
+    }
+    this.store[key][field] = value;
```
Comment (alanfriedman) [inline server/src/store/Memory.js:29]: For [Redis parity](https://redis.io/commands/hset)
```suggestion
    this.store[key][field] = value;
    return 1;
```

**Comment by jrmi** [comment]
> Thanks for this contribution; it looks great! clap
> 
> I pulled it locally and it works great. I would just suggest we keep the API consistent across stores as much as possible. We can adopt the Redis API for now and define our own API later, if needed.

Good idea for the API, job done !

(I don't know if in your workflow I must resolve your comments or if you do ?)

**Pull request merged**